### PR TITLE
Zero width space

### PR DESCRIPTION
**Pull request by NickPapacostas**
small change to fix a problem some people were experiencing when copy-pasting from emails to search. 

Zero-width-space is a valid UTF-8 character but because it's invisible can cause confusion. I know it's a particularly specific fix, but it seemed better than doing something fancy.

### CODE REVIEW

**Comment by jamesgorrie** [comment]
We could do it [here](https://github.com/guardian/grid/blob/master/kahuna/public/js/search/index.js#L41-L47) to keep this controller clean?

**Comment by NickPapacostas** [comment]
Ok so I realize that what i just pushed up might be substantial overkill, but i thought it might be more future-extensible or w.e 

If not, then I can either just add it to the existing matcher that was there (and maybe change it's name)
wanted to know what y'all thought

@jamesgorrie

File: kahuna/public/js/search/index.js [hunk 1]
```diff
@@ -35,15 +35,28 @@ export var search = angular.module('kahuna.search', [
...
-    // see: https://github.com/angular-ui/ui-router/issues/1119#issuecomment-64696060
     // TODO: Fix this
-    $urlMatcherFactoryProvider.type('Slashed', {
-        encode: val => angular.isDefined(val) ? val : undefined,
-        decode: val => angular.isDefined(val) ? val : undefined,
-        // We want to always match this type. If we match on slash, it won't update
-        // the `stateParams`.
-        is: () => true
+    $urlMatcherFactoryProvider.type('Query', {}, () => {
+        var queryFilters = {
```
Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: We try to use const where we can to avoid reassigning the var name.

File: kahuna/public/js/search/index.js [hunk 2]
```diff
@@ -35,15 +35,28 @@ export var search = angular.module('kahuna.search', [
...
+            // Query params slashes are encoded that mess up our seach query when we search for
+            // e.g. `24/7`.
+            // see: https://github.com/angular-ui/ui-router/issues/1119#issuecomment-64696060
+            defined: val => angular.isDefined(val) ? val : undefined,
+            // The string replacement is for removing the invisible "zero-width-space" UTF-8
+            // causing queries to fail: https://en.wikipedia.org/wiki/Zero-width_space
+            removeUtf8SpecialChars: val => val.replace(/%E2%80%8B/ig, '%20')
+        };
+        queryFilters.call = (val) => {
```
Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: we could just have this as a call function (to avoid mutating the obj). Not fussed, just a thought.

**Comment by jamesgorrie** [comment]
:+1: definitely easier to extend later! Nice.

File: kahuna/public/js/search/index.js (hunk 1 above)
Comment (NickPapacostas) [inline kahuna/public/js/search/index.js]: good call. Still adapting :)

File: kahuna/public/js/search/index.js [hunk 3]
```diff
@@ -35,15 +35,28 @@ export var search = angular.module('kahuna.search', [
...
-        is: () => true
+    $urlMatcherFactoryProvider.type('Query', {}, () => {
+        var queryFilters = {
+            // Query params slashes are encoded that mess up our seach query when we search for
+            // e.g. `24/7`.
+            // see: https://github.com/angular-ui/ui-router/issues/1119#issuecomment-64696060
+            defined: val => angular.isDefined(val) ? val : undefined,
+            // The string replacement is for removing the invisible "zero-width-space" UTF-8
+            // causing queries to fail: https://en.wikipedia.org/wiki/Zero-width_space
+            removeUtf8SpecialChars: val => val.replace(/%E2%80%8B/ig, '%20')
```
Comment (theefer) [inline kahuna/public/js/search/index.js]: If `defined` returned `undefined`, does this risk to throw?

**Comment by NickPapacostas** [comment]
@jamesgorrie we figured it out... kinda... basically "is" doesn't do what you think it does. If it looks good give me another thumbs up ^_^

**Comment by theefer** [comment]
Oh we forgot to convert the replace to use a regexp with the m flag.

File: kahuna/public/js/search/index.js [hunk 4]
```diff
@@ -35,15 +35,16 @@ export var search = angular.module('kahuna.search', [
...
+    const zeroWidthSpace = '\u200B';
+    function removeUtf8SpecialChars(val) {
+        return val && val.replace(zeroWidthSpace, '');
+    }
+
+    $urlMatcherFactoryProvider.type('Query', {
+        encode: val => removeUtf8SpecialChars(val),
+        decode: val => removeUtf8SpecialChars(val),
+        //call decode value that includes zero-width-space character
+        is: val => val && (val.indexOf(zeroWidthSpace) === -1)
```
Comment (jamesgorrie) [inline kahuna/public/js/search/index.js]: What now happens to values that aren't defined?

**Comment by jamesgorrie** [comment]
There seems to be a bug where, when you remove the last `char` from the query, it doesn't update the $stateParams. i.e. type `dog => backspace => backspace`. Now all you have is `d`. One more `backspace`, you still have `d` in the `qs`. I am assuming that's because we aren;t matching the `is` as `val`.

`is: angular.isDefined(val) && (val.indexOf(zeroWidthSpace) === -1)` seems to work as we aren't matching false on empty string then.

**Comment by theefer** [comment]
Good catch @jamesgorrie. Good reminder to do a full manual regression of the affected feature (search), esp around corner cases like empty input etc, when fiddling with the underlying router.

**Comment by NickPapacostas** [comment]
@jamesgorrie @theefer ty! Saw the bug, made the change, confirmed it worked :) 

I'll make sure to check thoroughly in the future, any chance there's a "critical path" kind of doc with stuff to check in a regression?

**Comment by theefer** [comment]
Sadly no. My test is usually: search, upload an image, edit its metadata, crop it.

In this instance it'd probably just involve doing a few corner case searches.

File: kahuna/public/js/search/index.js [hunk 5]
```diff
@@ -114,7 +115,7 @@ search.config(['$stateProvider', '$urlMatcherFactoryProvider',
                 ).
                     distinctUntilChanged(angular.identity, Immutable.is).
                     shareReplay(1);
-            }]
+            }],
```
Comment (theefer) [inline kahuna/public/js/search/index.js]: Unnecessary (does it pass lint?)

File: kahuna/public/js/search/index.js [hunk 6]
```diff
@@ -35,15 +35,16 @@ export var search = angular.module('kahuna.search', [
...
-    // TODO: Fix this
-    $urlMatcherFactoryProvider.type('Slashed', {
-        encode: val => angular.isDefined(val) ? val : undefined,
-        decode: val => angular.isDefined(val) ? val : undefined,
-        // We want to always match this type. If we match on slash, it won't update
-        // the `stateParams`.
-        is: () => true
+    const zeroWidthSpace = '\u200B';
+    function removeUtf8SpecialChars(val) {
+        return val && val.replace(zeroWidthSpace, '');
```
Comment (theefer) [inline kahuna/public/js/search/index.js]: This only replaces the first one. Did we want to be safe and replace all zero-width space?

**Comment by NickPapacostas** [comment]
ok cool ^_^ shall i take that as a thumbs up?

**Comment by jamesgorrie** [comment]
Could we do a global replace as per @theefer's [comment](https://github.com/guardian/grid/pull/1539#discussion_r47078616)?

**Comment by NickPapacostas** [comment]
@theefer @jamesgorrie Yep, sorry somehow missed those comments. Lint passed, regexp being used, manually tested with multiple special charz in there. Sorry for the sloppiness ^_^

**Comment by jamesgorrie** [comment]
Not sloppy at all, this is the third / fourth time it's being done as it's not a very clear API.
This will hopefully be the one :clap: :+1:

**Comment by theefer** [comment]
Great, don't forget to go and tell Mee-Lai!
